File: 2019 카카오 인턴십/1. 크레인 인형뽑기 게임.py
# https://programmers.co.kr/learn/courses/30/lessons/64061
# 1. 크레인 인형뽑기 게임
# 걸린 시간: 0:19

def solution(board, moves):
    answer = 0
    N = len(board)
    top = [0] * N
    stack = []

    for c in range(N):
        for r in range(N - 1, -1, -1):
            if board[r][c] == 0:
                top[c] = r + 1
                break

    for m in moves:
        m -= 1
        if top[m] < N:
            stack.append( board[top[m]][m] )
            top[m] += 1

            if len(stack) >= 2 and stack[-1] == stack[-2]:
                answer += 2
                stack.pop()
                stack.pop()

    return answer

board = [[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]]
moves = [1,5,3,5,1,2,1,4]
print( solution(board, moves) ) # 4

File: 2019 카카오 인턴십/1. 크레인 인형뽑기 게임.js
// https://programmers.co.kr/learn/courses/30/lessons/64061
// 1. 크레인 인형뽑기 게임

function solution(board, moves) {
    var answer = 0;
    var N = board.length;
    var lines = Array(N).fill().map(() => Array())
    var stack = Array();

    for (let c = 0; c < N; c++) {
        for (let r = N - 1; r >= 0; r--) {
            if (board[r][c] > 0)
                lines[c].push( board[r][c] );
            else break;
        }
    }

    for (let m of moves) {
        m -= 1;
        if (lines[m].length > 0) { 
            stack.push( lines[m].pop() )
            if (stack.length >= 2 && stack[stack.length - 1] == stack[stack.length - 2]) {
                answer += 2;
                stack.pop();
                stack.pop();
            }
        }
    }

    return answer;
}

board = [[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]]
moves = [1,5,3,5,1,2,1,4]
console.log( solution(board, moves) )